reconcile/utils/cluster_version_data.py

```diff
@@ -0,0 +1,54 @@
+from dataclasses import dataclass, asdict, field
+from typing import Any, Optional
+
+from reconcile.utils.state import State
+
+
+@dataclass
+class WorkloadHistory:
+    soak_days: float = 0.0
+    reporting: list[str] = field(default_factory=list)
+
+
+@dataclass
+class VersionHistory:
+    workloads: dict[str, WorkloadHistory]
+
+
+@dataclass
+class VersionData:
+    check_in: str
+    versions: dict[str, VersionHistory]
+
+    def workload_history(
+        self, version: str, workload: str, default: Optional[WorkloadHistory] = None
+    ) -> WorkloadHistory:
+        if not default:
+            vh = self.versions.get(version, VersionHistory({}))
+            return vh.workloads.get(workload, WorkloadHistory())
+        vh = self.versions.setdefault(version, VersionHistory({}))
+        return vh.workloads.setdefault(workload, default)
+
+    def asdict(self) -> dict[str, Any]:
+        return asdict(self)
+
+    def save(self, state: State, ocm_name: str) -> None:
+        state.add(ocm_name, asdict(self), force=True)
+
+
+def version_data_from_dict(d: dict[str, Any]) -> VersionData:
```

Comment:
this is the function that bastically restores `VersionData` from a json doc stored in `State` right?

if you use `pydantic` dataclasses over regular ones, you get object tree unmarshalling for free like

```python
from pydantic.dataclasses import dataclass

@dataclass
class VersionData:
   ...

vd = VersionData(**d)
```

Reply:
great, thanks! I will have a look.

Reply:
I went for Pydantic `BaseModel` instead of `dataclass` in order to more easily encode/decode datetime... and then I ended up just modifying the format to use the default iso format. I still think using `BaseModel` is fine though. Let me know what you think

Reply:
nice. yes both pydantic options help with the tree unmarshalling
thanks!